### PR TITLE
Proposal for CDATA handling.

### PR DESCRIPTION
**Pull request by dskatz**
This offers a potential solution for submitting CDATA text nodes in a SOAP Request.  

When integrating with many Enterprise level third party APIs CDATA text nodes are often used.

### CODE REVIEW

**Comment by herom** [comment]
Thanks a lot for your time & effort @dskatz - we appreciate it a lot!

According the tests I don't know if it would be more sufficient if they were done as request/response samples? (ping @jsdevel)

**Comment by jsdevel** [comment]
I think we can accept them as is.  Any chance you can squash commits down to 1 @dskatz ?

**Comment by dskatz** [comment]
That is not a problem.  I didn't see your latest message @jsdevel.  I'll revert to d330d9 and squash into one commit.

**Comment by jsdevel** [comment]
Looking forward to it @dskatz !

**Comment by paveytel** [comment]
Hello @jsdevel @dskatz , Quick question, when this is merged, will node-soap be able to process CDATA ?, currently node-soap formats my CDATA(a string embedded within a JS object) replacing the "<" and ">" tags with "&lt" and "&gt". Respectively. I just wish to know if this is as a result of nodesoap not handling CDATA, 
Thanks

**Comment by jsdevel** [comment]
> when this is merged, will node-soap be able to process CDATA ?, currently node-soap formats my CDATA(a string embedded within a JS object) replacing the "<" and ">" tags with "&lt" and "&gt". Respectively. I just wish to know if this is as a result of nodesoap not handling CDATA, 

That's a great question.  To my knowledge, `node-soap` has never intentionally handled CDATA.  I would expect that you'd want to just get a string back though, with no object wrapper around it.

File: test/request-response-samples/SendCDATA__cdata_is_handled_correctly/request.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "TestMessage":{
+        "message": "<![CDATA[Hello World! & Hello Me!]]>"
```
Comment (jsdevel) [inline test/request-response-samples/SendCDATA__cdata_is_handled_correctly/request.json]: should we strip `<![CDATA[]]>`?

Comment (paveytel) [inline test/request-response-samples/SendCDATA__cdata_is_handled_correctly/request.json]: this worked for me tho, adding this in lib/wsdl.js on line 69 to 71
```if (obj.substr(0,9) === '<![CDATA[' && obj.substr(-3) === "]]>") {
 +      return obj;
 +    }```

**Comment by herom** [comment]
ping @dskatz :smiley_cat:

File: test/request-response-samples/SendCDATA__cdata_is_handled_correctly/request.json (hunk 1 above)
Comment (stevezxu) [inline test/request-response-samples/SendCDATA__cdata_is_handled_correctly/request.json]: @jsdevel, <![CDATA[]]> should be leave as it is.

**Comment by stevezxu** [comment]
Any chance we could merge this pull request to master? Trying to use this as it is, being blocked by not  being able to handle CDATA. Thanks for great work. @dskatz @herom

**Comment by stevezxu** [comment]
Is there anything we can do to merge this pull request to master? @dakatz hasn't been replied for a month now. Thanks. @herom

**Comment by mspisars** [comment]
+1 on getting this into master...

File: test/request-response-samples/SendCDATA__cdata_is_handled_correctly/request.xml [hunk 2]
```diff
@@ -0,0 +1 @@
+<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"  xmlns:tns="http://example.com/testservice.wsdl" xmlns:xsd1="http://example.com/testservice.xsd"><soap:Body><xsd1:TestMessage xmlns:xsd1="http://example.com/testservice.xsd"><TestMessage><message><![CDATA[Hello World! & Hello Me!]]></message></TestMessage></xsd1:TestMessage></soap:Body></soap:Envelope>
```
Comment (aleung) [inline test/request-response-samples/SendCDATA__cdata_is_handled_correctly/request.xml]: Test failed because XML declaration is missing.

Comment (jsdevel) [inline test/request-response-samples/SendCDATA__cdata_is_handled_correctly/request.xml]: Thanks @aleung.  Can you add it?

**Comment by aleung** [comment]
@jsdevel I've submitted the fix.

**Pull request merged**